### PR TITLE
CORE-11759: Add support for Corda Prometheus PodMonitor

### PR DESCRIPTION
**Pull request by oscardalmau-r3**
We currently have an option exposed on the Corda Helm chart that adds Prometheus annotations to the workers pods. Promtail does not utilise these and, instead, the expectation is that Pod/ServiceMonitors are defined to identify endpoints to scrape. 

This PR adds support to create a PodMonitor. Unlike the annotations, this is disabled by default (as it would fail if the CRD is not defined in the environment).

### CODE REVIEW

File: charts/corda/values.yaml [hunk 1]
```diff
@@ -39,6 +39,13 @@ metrics:
   # -- enable scraping of worker metrics through Prometheus annotations
   scrape: true
 
+# Pod monitor configuration for scraping metrics. Please note that the corresponding CRD must exist in the environment
+podMonitor:
```
Comment (davidcurrie) [inline charts/corda/values.yaml]: This should be under `metrics`

File: charts/corda-lib/templates/_helpers.tpl [hunk 2]
```diff
@@ -513,6 +513,31 @@ data:
...
+{{/*
+Pod Monitor creation
+*/}}
+{{- define "corda.podMonitor" -}}
+{{- if .Values.podMonitor.enabled }}
+---
+apiVersion: monitoring.coreos.com/v1
+kind: PodMonitor
+metadata:
+  name: corda
```
Comment (davidcurrie) [inline charts/corda-lib/templates/_helpers.tpl]: the name should take into account the Helm release

File: charts/corda-lib/templates/_helpers.tpl [hunk 3]
```diff
@@ -513,6 +513,31 @@ data:
...
+  {{- range $k, $v := .Values.podMonitor.labels }}
+    {{ $k }}: {{ $v | quote }}
+  {{- end }}
+spec:
+  podMetricsEndpoints:
+  - port: monitor
+  jobLabel: corda
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: corda
```
Comment (davidcurrie) [inline charts/corda-lib/templates/_helpers.tpl]: this should use the `corda.name` function as we do when generating the label

File: charts/corda-lib/templates/_helpers.tpl [hunk 4]
```diff
@@ -513,6 +513,31 @@ data:
...
+metadata:
+  name: corda
+  labels:
+  {{- range $k, $v := .Values.podMonitor.labels }}
+    {{ $k }}: {{ $v | quote }}
+  {{- end }}
+spec:
+  podMetricsEndpoints:
+  - port: monitor
+  jobLabel: corda
```
Comment (davidcurrie) [inline charts/corda-lib/templates/_helpers.tpl]: can't remember how this is used but should it also use the release name?

File: charts/corda/values.schema.json [hunk 5]
```diff
@@ -188,6 +188,35 @@
...
+                    "default": false,
+                    "title": "enable pod monitor creation to identify endpoints to scrape",
+                    "examples": [
+                        false
+                    ]
+                },
+                "labels": {
+                    "type": "object",
+                    "default": {},
+                    "title": "labels that can be used so PodMonitor is discovered by Prometheus",
```
Comment (davidcurrie) [inline charts/corda/values.schema.json]: be consistent about "pod monitor" or "PodMonitor"

File: charts/corda/values.schema.json [hunk 6]
```diff
@@ -188,6 +188,35 @@
...
+                "enabled": {
+                    "type": "boolean",
+                    "default": false,
+                    "title": "enable pod monitor creation to identify endpoints to scrape",
+                    "examples": [
+                        false
+                    ]
+                },
+                "labels": {
+                    "type": "object",
```
Comment (davidcurrie) [inline charts/corda/values.schema.json:186]: can we be more precise about the properties of this object?

**Review by davidcurrie — CHANGES_REQUESTED** [review]

File: charts/corda/values.yaml (hunk 1 above)
Comment (oscardalmau-r3) [reply charts/corda/values.yaml]: Done

File: charts/corda-lib/templates/_helpers.tpl (hunk 2 above)
Comment (oscardalmau-r3) [reply charts/corda-lib/templates/_helpers.tpl]: Done - sorry I forgot about this

File: charts/corda-lib/templates/_helpers.tpl (hunk 3 above)
Comment (oscardalmau-r3) [reply charts/corda-lib/templates/_helpers.tpl]: Done

File: charts/corda-lib/templates/_helpers.tpl (hunk 4 above)
Comment (oscardalmau-r3) [reply charts/corda-lib/templates/_helpers.tpl]: Done

File: charts/corda/values.schema.json (hunk 5 above)
Comment (oscardalmau-r3) [reply charts/corda/values.schema.json]: Done

File: charts/corda/values.schema.json (hunk 6 above)
Comment (oscardalmau-r3) [reply charts/corda/values.schema.json:186]: @davidcurrie I don't think we can. In the `non-functional-tests` we use `release` key, but Prometheus can be installed in different ways and `release` may not be an option always.

**Review by davidcurrie — APPROVED** [review]
👍

**Pull request merged**